File: guilib-bukkit/src/main/java/net/developertobi/guilib/bukkit/api/gui/GuiControllerImpl.java
package net.developertobi.guilib.bukkit.api.gui;

import net.developertobi.guilib.api.GuiProvider;
import net.developertobi.guilib.api.gui.Gui;
import net.developertobi.guilib.api.gui.GuiController;
import net.developertobi.guilib.api.gui.GuiProperties;
import net.developertobi.guilib.api.item.GuiItem;
import net.developertobi.guilib.api.item.GuiPos;
import net.developertobi.guilib.api.pagination.GuiPagination;
import net.developertobi.guilib.api.utils.MathUtils;
import net.developertobi.guilib.bukkit.api.pagination.GuiPaginationImpl;
import net.kyori.adventure.text.serializer.plain.PlainTextComponentSerializer;
import org.bukkit.Material;
import org.bukkit.inventory.Inventory;

import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.ThreadLocalRandom;

public class GuiControllerImpl implements GuiController {

    private final Gui provider;
    private final GuiProperties properties;
    private final int slotCount;
    private boolean closeable;
    private final Map<GuiPos, GuiItem> contents = new HashMap<>();
    private GuiPagination pagination;
    private Inventory rawInventory;
    private String overriddenGuiId;
    private int overriddenRows;
    private int overriddenColumns;
    private GuiPos inputAreaStart;
    private GuiPos inputAreaEnd;

    public GuiControllerImpl(Gui provider) {
        this.provider = provider;
        GuiProperties ann = provider.getClass().getAnnotation(GuiProperties.class);
        if (ann == null) {
            throw new IllegalArgumentException("Gui provider must have @GuiProperties annotation: " + provider.getClass().getName());
        }
        this.properties = ann;
        this.slotCount = properties.rows() * properties.columns();
        this.closeable = properties.closeable();
        this.overriddenGuiId = null;
        this.overriddenRows = 0;
        this.overriddenColumns = 0;
    }

    @Override
    public Gui getProvider() {
        return provider;
    }

    @Override
    public GuiProperties getProperties() {
        return properties;
    }

    @Override
    public int getSlotCount() {
        return slotCount;
    }

    @Override
    public boolean isCloseable() {
        return closeable;
    }

    @Override
    public void setCloseable(boolean closeable) {
        this.closeable = closeable;
    }

    @Override
    public Map<GuiPos, GuiItem> getContents() {
        return contents;
    }

    @Override
    public GuiPagination getPagination() {
        return pagination;
    }

    public void setPagination(GuiPagination pagination) {
        this.pagination = pagination;
    }

    @Override
    public Inventory getRawInventory() {
        return rawInventory;
    }

    @Override
    public void setRawInventory(Inventory rawInventory) {
        this.rawInventory = rawInventory;
    }

    @Override
    public String getOverriddenGuiId() {
        return overriddenGuiId;
    }

    @Override
    public void setOverriddenGuiId(String overriddenGuiId) {
        this.overriddenGuiId = overriddenGuiId;
    }

    @Override
    public int getOverriddenRows() {
        return overriddenRows;
    }

    @Override
    public void setOverriddenRows(int overriddenRows) {
        this.overriddenRows = overriddenRows;
    }

    @Override
    public int getOverriddenColumns() {
        return overriddenColumns;
    }

    @Override
    public void setOverriddenColumns(int overriddenColumns) {
        this.overriddenColumns = overriddenColumns;
    }

    @Override
    public String getGuiId() {
        return overriddenGuiId != null ? overriddenGuiId : properties.id();
    }

    @Override
    public int getRows() {
        return overriddenRows != 0 ? overriddenRows : properties.rows();
    }

    @Override
    public int getColumns() {
        return overriddenColumns != 0 ? overriddenColumns : properties.columns();
    }

    @Override
    public void placeholder(GuiPos pos, Material type) {
        setItem(pos, GuiProvider.getApi().placeholder(type));
    }

    @Override
    public void placeholder(int row, int column, Material type) {
        setItem(row, column, GuiProvider.getApi().placeholder(type));
    }

    @Override
    public void setInputArea(GuiPos start, GuiPos end) {
        this.inputAreaStart = start;
        this.inputAreaEnd = end;
    }

    @Override
    public boolean isInInputArea(GuiPos pos) {
        if (inputAreaStart == null || inputAreaEnd == null) return false;
        int minRow = Math.min(inputAreaStart.row(), inputAreaEnd.row());
        int maxRow = Math.max(inputAreaStart.row(), inputAreaEnd.row());
        int minCol = Math.min(inputAreaStart.column(), inputAreaEnd.column());
        int maxCol = Math.max(inputAreaStart.column(), inputAreaEnd.column());
        return pos.row() >= minRow && pos.row() <= maxRow
                && pos.column() >= minCol && pos.column() <= maxCol;
    }

    @Override
    public void setItem(GuiPos pos, GuiItem item) {
        contents.put(pos, item);
    }

    @Override
    public void setItem(int row, int column, GuiItem item) {
        contents.put(GuiPos.of(row, column), item);
    }

    @Override
    public void addItem(GuiItem item) {
        GuiPos emptyPosition = getFirstEmptyPosition();
        if (emptyPosition == null) return;
        setItem(emptyPosition, item);
    }

    @Override
    public void addItemToRandomPosition(GuiItem item) {
        int randomSlotIndex = ThreadLocalRandom.current().nextInt(slotCount);
        GuiPos randomPosition = MathUtils.slotToPosition(randomSlotIndex, getColumns());
        if (isPositionTaken(randomPosition)) return;
        setItem(randomPosition, item);
    }

    @Override
    public void removeItem(String name) {
        PlainTextComponentSerializer serializer = PlainTextComponentSerializer.plainText();
        for (Map.Entry<GuiPos, GuiItem> entry : contents.entrySet()) {
            GuiItem guiItem = entry.getValue();
            if (guiItem == null) continue;

            if (!serializer.serialize(guiItem.getItem().displayName()).equalsIgnoreCase(name)) continue;

            contents.put(entry.getKey(), null);
            if (rawInventory != null) {
                rawInventory.remove(guiItem.getItem());
            }
        }
    }

    @Override
    public void removeItem(Material type) {
        for (Map.Entry<GuiPos, GuiItem> entry : contents.entrySet()) {
            GuiItem guiItem = entry.getValue();
            if (guiItem == null) continue;
            if (guiItem.getItem().getType() != type) continue;

            contents.put(entry.getKey(), null);
        }
    }

    @Override
    public void fill(FillType fillType, GuiItem item, GuiPos... positions) {
        int rows = getRows();
        int columns = getColumns();

        switch (fillType) {
            case ROW -> {
                if (positions.length > 1) {
                    throw new IllegalArgumentException("To fill a row only 1 position is allowed. Used positions: " + positions.length);
                }
                int startSlot = MathUtils.positionToSlot(positions[0], getColumns());
                for (int currentSlot = startSlot; currentSlot < startSlot + columns; currentSlot++) {
                    setItem(MathUtils.slotToPosition(currentSlot, columns), item);
                }
            }

            case RECTANGLE -> {
                if (positions.length != 2) {
                    throw new IllegalArgumentException("Only two positions are allowed to create a rectangle!");
                }
                GuiPos fromPos = positions[0];
                GuiPos toPos = positions[1];
                int fromRow = fromPos.row();
                int fromColumn = fromPos.column();
                int toRow = toPos.row();
                int toColumn = toPos.column();

                for (int row = fromRow; row <= toRow; row++) {
                    for (int col = fromColumn; col <= toColumn; col++) {
                        setItem(GuiPos.of(row, col), item);
                    }
                }
            }

            case LEFT_BORDER -> {
                int currentSlot = 0;
                while (currentSlot < slotCount) {
                    GuiPos currentPosition = MathUtils.slotToPosition(currentSlot, columns);
                    setItem(currentPosition, item);
                    currentSlot += rows;
                }
            }

            case RIGHT_BORDER -> {
                int lastColumnStart = rows - 1;
                int lastColumnEnd = slotCount - 1;
                int currentSlot = lastColumnStart;
                while (currentSlot <= lastColumnEnd) {
                    GuiPos currentPos = MathUtils.slotToPosition(currentSlot, columns);
                    GuiPos nextPos = MathUtils.nextPositionFromSlot(currentSlot, columns);

                    if (currentPos.row() == nextPos.row()) {
                        currentSlot += rows;
                        continue;
                    }

                    GuiPos currentPosition = MathUtils.slotToPosition(currentSlot, columns);
                    setItem(currentPosition, item);
                    currentSlot += rows;
                }
            }

            case TOP_BORDER -> {
                for (int currentSlot = 0; currentSlot < columns; currentSlot++) {
                    GuiPos currentPosition = MathUtils.slotToPosition(currentSlot, columns);
                    setItem(currentPosition, item);
                }
            }

            case BOTTOM_BORDER -> {
                int size = slotCount;
                int firstColumnInLastRow = size - columns;
                for (int currentSlot = firstColumnInLastRow; currentSlot < size; currentSlot++) {
                    GuiPos currentPosition = MathUtils.slotToPosition(currentSlot, columns);
                    setItem(currentPosition, item);
                }
            }

            case ALL_BORDERS -> {
                fill(FillType.TOP_BORDER, item);
                fill(FillType.RIGHT_BORDER, item);
                fill(FillType.BOTTOM_BORDER, item);
                fill(FillType.LEFT_BORDER, item);
            }
        }
    }

    @Override
    public void clearPosition(GuiPos pos) {
        contents.put(pos, null);
        if (rawInventory != null) {
            rawInventory.clear(MathUtils.positionToSlot(pos, getColumns()));
        }
    }

    @Override
    public boolean isPositionTaken(GuiPos pos) {
        return contents.get(pos) != null;
    }

    @Override
    public GuiPos getPositionOfItem(GuiItem item) {
        for (Map.Entry<GuiPos, GuiItem> entry : contents.entrySet()) {
            if (entry.getValue() == item) {
                return entry.getKey();
            }
        }
        return null;
    }

    @Override
    public GuiPos getFirstEmptyPosition() {
        int columns = getColumns();
        for (int i = 0; i < slotCount; i++) {
            GuiPos position = MathUtils.slotToPosition(i, columns);
            if (contents.get(position) == null) {
                return position;
            }
        }
        return null;
    }

    @Override
    public GuiItem getItem(GuiPos pos) {
        return contents.get(pos);
    }

    @Override
    public GuiItem getItem(int row, int column) {
        return contents.get(GuiPos.of(row, column));
    }

    @Override
    public GuiItem findFirstItemWithType(Material type) {
        for (int slot = 0; slot < slotCount; slot++) {
            GuiPos currentPosition = MathUtils.slotToPosition(slot, getColumns());
            GuiItem guiItem = contents.get(currentPosition);
            if (guiItem == null) continue;
            if (guiItem.getItem() == null) continue;
            if (guiItem.getItem().getType() != type) continue;
            return guiItem;
        }
        return null;
    }

    @Override
    public GuiPagination createPagination() {
        if (pagination == null) {
            pagination = new GuiPaginationImpl(this);
        }
        return pagination;
    }

    @Override
    public void updateRawInventory() {
        for (Map.Entry<GuiPos, GuiItem> entry : contents.entrySet()) {
            GuiItem guiItem = entry.getValue();
            if (guiItem == null) continue;
            if (rawInventory != null) {
                rawInventory.setItem(MathUtils.positionToSlot(entry.getKey(), getColumns()), guiItem.getItem());
            }
        }
    }
}

File: guilib-bukkit/bin/main/net/developertobi/inventorylib/bukkit/api/inventory/GuiControllerImpl.kt
package net.developertobi.inventorylib.bukkit.api.inventory

import net.kyori.adventure.text.serializer.plain.PlainTextComponentSerializer
import net.developertobi.inventorylib.api.inventory.GuiController
import net.developertobi.inventorylib.api.inventory.GuiProperties
import net.developertobi.inventorylib.api.inventory.Gui
import net.developertobi.inventorylib.api.item.GuiItem
import net.developertobi.inventorylib.api.item.GuiPos
import net.developertobi.inventorylib.api.pagination.GuiPagination
import net.developertobi.inventorylib.api.utils.MathUtils
import net.developertobi.inventorylib.bukkit.api.pagination.GuiPaginationImpl
import net.developertobi.inventorylib.api.GuiProvider
import org.bukkit.Material
import org.bukkit.inventory.Inventory
import java.util.concurrent.ThreadLocalRandom

class GuiControllerImpl(override val provider: Gui) : GuiController {

    override val properties: GuiProperties = provider.javaClass.getAnnotation(GuiProperties::class.java)

    override val inventorySlotCount: Int = properties.rows * properties.columns
    override var isCloseable: Boolean = properties.closeable

    override val contents: MutableMap<GuiPos, GuiItem?> = mutableMapOf()
    override var pagination: GuiPagination? = null
    override var rawInventory: Inventory? = null

    override var overriddenInventoryId: String? = null
    override var overriddenRows: Int = 0
    override var overriddenColumns: Int = 0

    override fun getInventoryId(): String {
        return if (overriddenInventoryId == null) this.properties.id else overriddenInventoryId!!
    }

    override fun getRows(): Int {
        return if (overriddenRows == 0) this.properties.rows else overriddenRows
    }

    override fun getColumns(): Int {
        return if (overriddenColumns == 0) this.properties.columns else overriddenColumns
    }

    override fun placeholder(pos: GuiPos, type: Material) {
        setItem(pos, GuiProvider.api.placeholder(type))
    }

    override fun placeholder(row: Int, column: Int, type: Material) {
        setItem(row, column, GuiProvider.api.placeholder(type))
    }

    override fun setItem(pos: GuiPos, item: GuiItem) {
        contents.replace(pos, item)
    }

    override fun setItem(row: Int, column: Int, item: GuiItem) {
        contents.replace(GuiPos.of(row, column), item)
    }

    override fun addItem(item: GuiItem) {
        val emptyPosition: GuiPos = getFirstEmptyPosition() ?: return
        setItem(emptyPosition, item)
    }

    override fun addItemToRandomPosition(item: GuiItem) {
        val randomSlotIndex = ThreadLocalRandom.current().nextInt(inventorySlotCount)
        val randomPosition: GuiPos = MathUtils.slotToPosition(randomSlotIndex, getColumns())
        if (isPositionTaken(randomPosition)) return
        setItem(randomPosition, item)
    }

    override fun removeItem(name: String) {
        val tempEntries: Set<Map.Entry<GuiPos, GuiItem?>> =
            contents.entries
        for ((position, value) in tempEntries) {
            val guiItem: GuiItem = value ?: continue

            val serializer: PlainTextComponentSerializer = PlainTextComponentSerializer.plainText()
            if (!serializer.serialize(guiItem.item.displayName()).equals(name, true)) continue

            contents.replace(position, null)
            this.rawInventory?.remove(guiItem.item)
        }
    }

    override fun removeItem(type: Material) {
        val tempEntries: Set<Map.Entry<GuiPos, GuiItem?>> = contents.entries

        for ((position, value) in tempEntries) {
            val guiItem: GuiItem = value ?: return
            if (guiItem.item.type != type) continue

            contents.replace(position, null)
        }
    }

    override fun fill(fillType: GuiController.FillType, item: GuiItem, vararg positions: GuiPos) {
        val rows: Int = getRows()
        val columns: Int = getColumns()

        when (fillType) {
            GuiController.FillType.ROW -> {
                require(positions.size <= 1) { "To fill a row only 1 position is allowed. Used positions: " + positions.size }

                val startSlot: Int = MathUtils.positionToSlot(positions[0], getColumns())

                for (currentSlot in startSlot until (startSlot + columns)) {
                    setItem(MathUtils.slotToPosition(currentSlot, columns), item)
                }
            }

            GuiController.FillType.RECTANGLE -> {
                require(positions.size == 2) { "Only two positions are allowed to create a rectangle!" }

                val fromPos: GuiPos = positions[0]
                val toPos: GuiPos = positions[1]

                val fromRow: Int = fromPos.row
                val fromColumn: Int = fromPos.column
                val toRow: Int = toPos.row
                val toColumn: Int = toPos.column

                for (row in fromRow..toRow) {
                    for (col in fromColumn..toColumn) {
                        setItem(GuiPos.of(row, col), item)
                    }
                }
            }

            GuiController.FillType.LEFT_BORDER -> {
                var currentSlot = 0
                while (currentSlot < this.inventorySlotCount) {
                    val currentPosition: GuiPos = MathUtils.slotToPosition(currentSlot, columns)
                    setItem(currentPosition, item)
                    currentSlot += rows
                }
            }

            GuiController.FillType.RIGHT_BORDER -> {
                val lastColumnStart = rows - 1
                val lastColumnEnd = this.inventorySlotCount - 1
                var currentSlot = lastColumnStart
                while (currentSlot <= lastColumnEnd) {
                    val currentPos: GuiPos = MathUtils.slotToPosition(currentSlot, columns)
                    val nextPos: GuiPos = MathUtils.nextPositionFromSlot(currentSlot, columns)

                    if (currentPos.row == nextPos.row) {
                        currentSlot += rows
                        continue
                    }

                    val currentPosition: GuiPos = MathUtils.slotToPosition(currentSlot, columns)
                    setItem(currentPosition, item)
                    currentSlot += rows
                }
            }

            GuiController.FillType.TOP_BORDER -> {
                for (currentSlot in 0 until columns) {
                    val currentPosition: GuiPos = MathUtils.slotToPosition(currentSlot, columns)
                    setItem(currentPosition, item)
                }
            }

            GuiController.FillType.BOTTOM_BORDER -> {
                val size = this.inventorySlotCount
                val firstColumnInLastRow = size - columns
                for (currentSlot in firstColumnInLastRow until size) {
                    val currentPosition: GuiPos = MathUtils.slotToPosition(currentSlot, columns)
                    setItem(currentPosition, item)
                }
            }

            GuiController.FillType.ALL_BORDERS -> {
                fill(GuiController.FillType.TOP_BORDER, item)
                fill(GuiController.FillType.RIGHT_BORDER, item)
                fill(GuiController.FillType.BOTTOM_BORDER, item)
                fill(GuiController.FillType.LEFT_BORDER, item)
            }
        }
    }

    override fun clearPosition(pos: GuiPos) {
        contents.replace(pos, null)
        rawInventory?.clear(MathUtils.positionToSlot(pos, getColumns()))
    }

    override fun isPositionTaken(pos: GuiPos): Boolean {
        return contents[pos] != null
    }

    override fun getPositionOfItem(item: GuiItem): GuiPos? {
        return contents.entries.filter { it.value == item }.map { it.key }.firstOrNull()
    }

    override fun getFirstEmptyPosition(): GuiPos? {
        var emptyPosition: GuiPos? = null

        for (position in this.contents.keys) if (this.contents[position] == null) {
            emptyPosition = position
            break
        }

        return emptyPosition
    }

    override fun getItem(pos: GuiPos): GuiItem? {
        return contents[pos]
    }

    override fun getItem(row: Int, column: Int): GuiItem? {
        return contents[GuiPos.of(row, column)]
    }

    override fun findFirstItemWithType(type: Material): GuiItem? {
        var result: GuiItem? = null

        for (slot in 0..inventorySlotCount) {
            val currentPosition: GuiPos = MathUtils.slotToPosition(slot, getColumns())
            if (contents[currentPosition] == null) continue
            if (contents[currentPosition]?.item == null) continue
            if (contents[currentPosition]?.item?.type !== type) continue
            result = contents[currentPosition]
        }

        return result
    }

    override fun createPagination(): GuiPagination {
        if (this.pagination == null) this.pagination = GuiPaginationImpl(this)
        return this.pagination!!
    }

    override fun updateRawInventory() {
        for ((position, value) in this.contents) {
            val guiItem: GuiItem = value ?: continue
            rawInventory?.setItem(MathUtils.positionToSlot(position, getColumns()), guiItem.item)
        }
    }

}

